sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -339,11 +339,34 @@ case class DataSource(
         dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
       case (_: SchemaRelationProvider, None) =>
         throw new AnalysisException(s"A schema needs to be specified when using $className.")
-      case (dataSource: RelationProvider, Some(schema)) =>
+      case (dataSource: RelationProvider, Some(specifiedSchema)) =>
         val baseRelation =
           dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
-        if (baseRelation.schema != schema) {
-          throw new AnalysisException(s"$className does not allow user-specified schemas.")
+        val persistentSchema = baseRelation.schema
+        val persistentSize = persistentSchema.size
+        val specifiedSize = specifiedSchema.size
+        if (persistentSize == specifiedSize) {
+          val (persistentFields, specifiedFields) = persistentSchema.zip(specifiedSchema)
```

Comment:
I think this isn't going to handle nested schema case.

If we're going to improve such error message case across the codebase, we might also think about having a common method (maybe something called `assertEquality` in `StructType`?) that checks each type recursively and shows a better message. Can we at least have a private method for this case in the future?